File: chatmock/upstream_errors.py
from __future__ import annotations

import re
from typing import Any, overload

_TRUNCATION_MARKER = "...(truncated)"
_MAX_SNIPPET_CHARS = 160
_MAX_MESSAGE_CHARS = 500
_REDACTION_PATTERNS: tuple[tuple[re.Pattern[str], str], ...] = (
    (re.compile(r"Authorization:\s*Bearer\s+[^\s,;]+", re.IGNORECASE), "Authorization: Bearer [redacted]"),
    (re.compile(r"\bBearer\s+(?!\[redacted\]\b)[^\s,;]+", re.IGNORECASE), "Bearer [redacted]"),
    (re.compile(r"\bsk-[A-Za-z0-9_-]+\b"), "[redacted]"),
    (re.compile(r"\b(session_id|access_token|token)=([^&\s,;]+)", re.IGNORECASE), r"\1=[redacted]"),
)


def build_upstream_error(
    message: Any,
    *,
    status_code: Any = None,
    body: Any = None,
    content_type: Any = None,
    exception: Any = None,
    phase: Any = None,
) -> dict[str, dict[str, str]]:
    base_message = _normalize_message(message)
    details: list[str] = []

    phase_text = _normalize_value(phase)
    if phase_text:
        details.append(f"phase={phase_text}")

    status_text = _normalize_value(status_code)
    if status_text:
        details.append(f"upstream_status={status_text}")

    content_type_text = _normalize_value(content_type)
    if content_type_text:
        details.append(f"content_type={content_type_text}")

    exception_text = _normalize_exception(exception)
    if exception_text:
        details.append(f"exception={exception_text}")

    body_text = _normalize_body(body)
    if body_text:
        details.append(f"body={body_text}")

    full_message = base_message
    if details:
        full_message = f"{base_message} ({'; '.join(details)})"

    return {"error": {"message": _truncate(full_message, _MAX_MESSAGE_CHARS)}}


def _normalize_message(message: Any) -> str:
    value = _normalize_value(message)
    if not value:
        return "Upstream error"
    return value


def _normalize_exception(exception: Any) -> str | None:
    if exception is None:
        return None
    if isinstance(exception, BaseException):
        detail = _normalize_value(exception)
        name = exception.__class__.__name__
        if detail and detail != name:
            return _truncate(f"{name}: {detail}", _MAX_SNIPPET_CHARS)
        return name
    return _truncate(_normalize_value(exception), _MAX_SNIPPET_CHARS)


def _normalize_body(body: Any) -> str | None:
    if body is None:
        return None
    return _truncate(_normalize_value(body), _MAX_SNIPPET_CHARS)


def _normalize_value(value: Any) -> str | None:
    if value is None:
        return None
    if isinstance(value, bytes):
        text = value.decode("utf-8", errors="ignore")
    elif isinstance(value, str):
        text = value
    else:
        text = _safe_string(value)
    text = _redact(text).strip()
    if not text:
        return None
    return text


def _safe_string(value: Any) -> str:
    try:
        return str(value)
    except Exception:
        try:
            name = value.__class__.__name__
        except Exception:
            name = "object"
        return f"<unprintable {name}>"


def _redact(text: str) -> str:
    redacted = text
    for pattern, replacement in _REDACTION_PATTERNS:
        redacted = pattern.sub(replacement, redacted)
    return redacted


@overload
def _truncate(value: str, limit: int) -> str: ...


@overload
def _truncate(value: None, limit: int) -> None: ...


def _truncate(value: str | None, limit: int) -> str | None:
    if value is None:
        return None
    if len(value) <= limit:
        return value
    return value[: limit - len(_TRUNCATION_MARKER)] + _TRUNCATION_MARKER

File: chatmock-rs/src/upstream_errors.rs
use std::sync::OnceLock;

use regex::Regex;

use crate::protocol::{ErrorPayload, OpenAiError};

const TRUNCATION_MARKER: &str = "...(truncated)";
const MAX_SNIPPET_CHARS: usize = 160;
const MAX_MESSAGE_CHARS: usize = 500;

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct UpstreamErrorContext {
    pub status_code: Option<u16>,
    pub body: Option<Vec<u8>>,
    pub content_type: Option<String>,
    pub exception: Option<String>,
    pub phase: Option<String>,
}

pub fn build_upstream_error(message: Option<&str>, context: UpstreamErrorContext) -> ErrorPayload {
    let base_message = normalize_text(message).unwrap_or_else(|| "Upstream error".to_string());
    let mut details = Vec::new();

    if let Some(phase) = normalize_owned(context.phase) {
        details.push(format!("phase={phase}"));
    }
    if let Some(status_code) = context.status_code {
        details.push(format!("upstream_status={status_code}"));
    }
    if let Some(content_type) = normalize_owned(context.content_type) {
        details.push(format!("content_type={content_type}"));
    }
    if let Some(exception) =
        normalize_owned(context.exception).map(|text| truncate(&text, MAX_SNIPPET_CHARS))
    {
        details.push(format!("exception={exception}"));
    }
    if let Some(body) = context
        .body
        .map(|body| String::from_utf8_lossy(&body).into_owned())
        .and_then(|body| normalize_owned(Some(body)))
        .map(|text| truncate(&text, MAX_SNIPPET_CHARS))
    {
        details.push(format!("body={body}"));
    }

    let mut full_message = base_message;
    if !details.is_empty() {
        full_message.push_str(" (");
        full_message.push_str(&details.join("; "));
        full_message.push(')');
    }

    ErrorPayload {
        error: OpenAiError {
            message: truncate(&full_message, MAX_MESSAGE_CHARS),
        },
    }
}

fn normalize_text(text: Option<&str>) -> Option<String> {
    normalize_owned(text.map(str::to_string))
}

fn normalize_owned(text: Option<String>) -> Option<String> {
    let text = text?;
    let redacted = redact(&text).trim().to_string();
    if redacted.is_empty() {
        return None;
    }
    Some(redacted)
}

fn redact(text: &str) -> String {
    let mut redacted = text.to_string();
    for redaction in redaction_patterns() {
        redacted = match redaction {
            RedactionPattern::Static(pattern, replacement) => {
                pattern.replace_all(&redacted, *replacement).into_owned()
            }
            RedactionPattern::Bearer(pattern) => pattern
                .replace_all(&redacted, |captures: &regex::Captures<'_>| {
                    let token = captures
                        .get(1)
                        .map(|capture| capture.as_str())
                        .unwrap_or_default();
                    if token.eq_ignore_ascii_case("[redacted]") {
                        captures
                            .get(0)
                            .map(|capture| capture.as_str())
                            .unwrap_or_default()
                            .to_string()
                    } else {
                        "Bearer [redacted]".to_string()
                    }
                })
                .into_owned(),
        };
    }
    redacted
}

fn truncate(value: &str, limit: usize) -> String {
    let char_count = value.chars().count();
    if char_count <= limit {
        return value.to_string();
    }
    let keep = limit.saturating_sub(TRUNCATION_MARKER.chars().count());
    value.chars().take(keep).collect::<String>() + TRUNCATION_MARKER
}

enum RedactionPattern {
    Static(Regex, &'static str),
    Bearer(Regex),
}

fn redaction_patterns() -> &'static [RedactionPattern] {
    static PATTERNS: OnceLock<Vec<RedactionPattern>> = OnceLock::new();
    PATTERNS
        .get_or_init(|| {
            vec![
                RedactionPattern::Static(
                    Regex::new(r"Authorization:\s*Bearer\s+[^\s,;]+").expect("authorization regex"),
                    "Authorization: Bearer [redacted]",
                ),
                RedactionPattern::Bearer(
                    Regex::new(r"\bBearer\s+([^\s,;]+)").expect("bearer regex"),
                ),
                RedactionPattern::Static(
                    Regex::new(r"\bsk-[A-Za-z0-9_-]+\b").expect("sk token regex"),
                    "[redacted]",
                ),
                RedactionPattern::Static(
                    Regex::new(r"\b(session_id|access_token|token)=([^&\s,;]+)")
                        .expect("key value token regex"),
                    "$1=[redacted]",
                ),
            ]
        })
        .as_slice()
}
